test/Microsoft.OpenApi.Readers.Tests/V2Tests/OpenApiDocumentTests.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Microsoft.OpenApi.Exceptions;
+using Microsoft.OpenApi.Models;
+using Microsoft.OpenApi.Readers.Exceptions;
+using Xunit;
+
+namespace Microsoft.OpenApi.Readers.Tests.V2Tests
+{
+    public class OpenApiDocumentTests
+    {
+        [Fact]
+        public void ShouldThrowWhenReferenceTypeIsInvalid()
+        {
+            var input = @"
+swagger: 2.0
+info: 
+  title: test
+  version: 1.0.0
+paths: 
+  '/':
+    get:
+      responses:
+        '200':
+          description: ok
+          schema:
+            $ref: '#/definition/does/notexist'
+";
+
+            var reader = new OpenApiStringReader(new OpenApiReaderSettings() {
+            });
+
+            OpenApiDocument doc = null;
+
+            Assert.Throws<OpenApiReaderException>(() =>
+            doc = reader.Read(input, out var diagnostic));
+
+            Assert.Null(doc);
+        }
+
+
+        [Fact]
+        public void ShouldThrowWhenReferenceDoesNotExist()
+        {
+            var input = @"
+swagger: 2.0
+info: 
+  title: test
+  version: 1.0.0
+paths: 
+  '/':
+    get:
+      produces: ['application/json']
+      responses:
+        '200':
+          description: ok
+          schema:
+            $ref: '#/definitions/doesnotexist'
+";
+
+            var reader = new OpenApiStringReader(new OpenApiReaderSettings()
+            {
+            });
+
+            OpenApiDocument doc = null;
+
+            doc = reader.Read(input, out var diagnostic);
+
+            diagnostic.Errors.ShouldBeEquivalentTo(new List<OpenApiError> {
+                new OpenApiError( new OpenApiException("Invalid Reference identifier 'doesnotexist'.")) });
+        }
+    }
```

Comment:
Why should we expect one case to throw an exception and the other case to simply add the error to the diagnostic? Shouldn't we standardize the behavior?

Reply:
Yes.  This is a problem because one error is raised from OpenAPI.Net and so is a OpenApiException which is caught.  However, the other raises OpenApiReaderException which isn't caught.  This is fixed in the next PR so both types of exceptions are caught and put into the diagnostics error collection.  I will fix the test when I merge.

---
In reply to: [194478408](https://github.com/Microsoft/OpenAPI.NET/pull/275#discussion_r194478408) [](ancestors = 194478408)